### DISCUSSION TITLE
TX81Z conditional processing based on received sysex

### DISCUSSION THREAD

**Discussion by Youngwhisk** [RomanKubiak/ctrlr#332] [General]
Hi,
I have a Yamaha TX81Z that I have been playing around with in ctrlr (not sound programming). It is about 2 feet away so I don't really need to do this part but it has become a challenge as I have not been able to get this to work as I want it.
All I am trying to do so far is have a single button that when clicked will tell me what the receive channel setting is. The TX81Z will only respond to sysex dumps if the proper receive channel is included in the dump request. That part is no problem. I can do a for loop that increments the channel and the dump is received once the proper channel is sent.  The problem is if the TX81Z is in omni mode I get 16 dumps back. The receive channel is received as byte 17 of the dump. My thoughts were to send a dump request on channel 1 only. If I get a reply, byte 17 will either be 00 if channel 1 or 0F if omni. That works fine. If I don't get a reply then I want to send the request on channels 2 to 16 to get a single response showing the correct channel. My problem is that I am trying to halt the code if 00 or 0F is received as I have the info I am looking for. However I don't seem to be able to break the code after 00 of 0F is received and the for loop continues giving me 15 more messages (if the unit is in omni). If I run all this from a button click module. I can determine what byte 17 is and put it into a global variable as a number in the midimessage received module, but my If statment back in the click module doesn't react to that variable. I thought it might be due to timing  and have tried a timer but while it runs I still receive too many messages. 
Perhaps I don't understand the flow in lua. If commands are running in the click module and a reply is processed in the midi message module, do they run at the same time or can I halt the processing in the click module until a condition is or isn't met in the midimessage module?

Maybe someone has a better concept of how to do this. I have so far spent way more time than this is worth, but I want to figure it out.

**Comment by Ted-juh** [comment]
Could you post the code with the for loop and the break statement? Or provide the panel ?

**Reply by Youngwhisk** [reply]
I have tried a lot of different things. I will consolidate it and post on Sunday.

**Reply by Ted-juh** [reply]
On a second note. Omni is what it is. "All channels". So if you send one CTRLR message to request a dump, the TX81Z will happily send 16 back, for free, if it's set to Omni Mode.

In other words, there is no way to "break" or "stop" the synth from sending messages if it's set to Omni Mode. It just does because it passes on that one CTRLR message to all channels. Unfortunately. 

What you could do?  At panel loaded: Request a system dump ( if that is where the midi transmitting or receiving channel is set ) and make an indicator (label, LEDs,  custom component, whatever) which tells you at which channel you are or if it's in Omni Mode. So you don't have to do a request later on. 

(I use a combobox which takes the setting when the panel is loaded and which I can change myself if needed.)

*Edited because of grammar mistakes and so on.

**Reply by Youngwhisk** [reply]
That was exactly what I wanted to do including using a combo box. I can probably ignore the extra messages and not worry about it but it just bothers me.
Actually if you send a dump request on a single channel (even in omni mode) it only responds with one dump which indicates the channel. The settings are 0-16, where 0-15 are for channels 1-16 and 16 is for omni. but to get the 0-15 you have to send on that channel. But if you send 16 then you get 16 replies.
I will play with it more and see what happens. I appreciate your help and have seen where you have helped others.

**Reply by Ted-juh** [reply]
So, no matter which channel you send the dump request to, it will always send a dump indicating which channel receives midi or whether it's in Omni Mode?

Then why the for loop to send 16 messages? I'm confused.

**Reply by Youngwhisk** [reply]
No,
If it is in omni you can send a dump request to any single channel and the single reply will indicate 0F.

If it is not omni, you have to send a dump request to the channel that is currently set as the receive channel. Seeing as I don't know what that is (supposedly), I use a for loop to step through each channel. There will only be a response from the correct channel. And that does work.

The problem is if it is in omni and I step through them all, I get a response from each channel.

**Reply by damiensellier** [reply]
@DnaldeT is right, you won't know which channel it is in omni. You should run a timer script scanning from ch0 to ch15 and wait for a reply. 
I don't know if it could help, you must check before, but yamaha products send replies to sysex general identity requests.
Supposedly the reply contains the current sysex id for operation and the manufacturer ID etc.
I use it whenever i can to check if the unit is online or not before going further on my panels.

f0 7e 7f 06 01 f7 -- Request Universal Sysex Device Info

Just check if you get a reply with that on the channel you're checking.

http://midi.teragonaudio.com/tech/midispec/identity.htm

**Reply by Ted-juh** [reply]
> The problem is if it is in Omni and I step through them all, I get a response from each channel.

It's not a problem. As said before it is what Omni is supposed to do. It receives on "all" channels. A for loop on all 16 channels means you'll receive 16 dumps. 

You should start without the "for" loop. 
-- Check whether the "receiving channel" is set to Channel 1 OR Omni. By just sending one Ctrlr message.
-- If you get 0F on Byte 17, it will tell you to not take any further action. 
-- If it doesn't return any response, it will tell you it is not Channel 1. (Byte 17 == nil)
-- Now you can do a "For" loop to check which channel responds. Again checking against Byte 17 if it returns a value or nil.

What kind of Dump are you requesting: System Setup Bulk Data Format (SYS) or Performance Edit Parameters (PCED)?

**Reply by dnaldoog** [reply]
> ... yamaha products send replies to sysex general identity requests.
> Supposedly the reply contains the current sysex id for operation and the manufacturer ID etc.
> I use it whenever i can to check if the unit is online or not before going further on my panels.
> 
> f0 7e 7f 06 01 f7 -- Request Universal Sysex Device Info
> 
> Just check if you get a reply with that on the channel you're checking.
> 
> http://midi.teragonaudio.com/tech/midispec/identity.htm

I didn't know that - Sure enough in the Korg MS2000 documentation:

| F0 | Exclusive Status |
| 7E | Non Realtime Message |
| nn | MIDI Channel (Device ID) |
| 06 | General Information |
| 01 | Identity Request |
| F7 | END OF EXCLUSIVE 

Thanks Damien.

**Reply by Youngwhisk** [reply]
__You should start without the "for" loop.
-- Check whether the "receiving channel" is set to Channel 1 OR Omni. By just sending one Ctrlr message.
-- If you get 0F on Byte 17, it will tell you to not take any further action.
-- If it doesn't return any response, it will tell you it is not Channel 1. (Byte 17 == nil)
-- Now you can do a "For" loop to check which channel responds. Again checking against Byte 17 if it returns a value or nil._

That's what I said in my first post.
send a message on channel 1.
1) receive 00, then receive channel is chan 1
2) receive 0F, then receive channel is omni
3) receive no reply, then receive channel is 2 - 16 and not omni. Use for loop through the remaining channels. 

The problem I am having is getting the code to stop running if the responce is 0F and it still loops through all the channels anyway.
I have been converting from Hex so I will try without the conversion. 

I have been using the System Setup (SYS=0) Dump request. It works as it should.

I will try the general midi request. That looks very promising and makes sense but it doesn't seem to be in the Sysex info (i may have missed it)

**Reply by damiensellier** [reply]
If it works, the universal sysex identity request is the shortest/fastest you can get providing the shortest/fastest reply as well. if it doesn't work, use your system setup bulk dump request and set the timer interval longer.

So, you have to create a script with a timer (leave 100ms between messages):

```
midiChannelTxCount = 0 -- for timer count
timer:setCallback (1, TimerCallback)
timer:startTimer(1, 100) -- 100ms cycle
```


Run a timer loop scanning all channels with your message whose variable is the midi channel byte value.
in another script called TimerCallback() : 

```
TimerCallback = function(timerId)

	if timerId == 1 and midiChannelTxCount <= 16 then
		console("midiChannelTxCount is : "..(midiChannelTxCount))
		panel:sendMidiMessageNow(CtrlrMidiMessage({0xf0, midiChannelTxCount, 0x00, 0x06, 0x01, 0xf7})) -- Request Universal Sysex Device Info
		midiChannelTxCount = midiChannelTxCount + 1

	else 
		timer:stopTimer(timerId) -- stop the timer loop
		console("timer 1 reset")
		midiChannelTxCount = 0 -- reset the counter
	end

end -- end timer script
```

I wrote it fast maybe there's some errors. Anyway it works like this: 
1. the timer is triggered by the function timer:setCallback() 
2. the loop starts with the midi message sent on channel 0
3. 100ms after the midi message is sent on channel 1 
4. and so on
5. if the proper message is received the loop is stopped and your midiChannel value is stored as "midiChannelTx"


in your midiMessageReceived script, add a statement like this (midiMessageReceived script needs to be selected in the global pannel properties) : 

```
midiMessageReceived = function(midiMessage, source)
midiChannelTxByte = midiMessage:getData():getByte(2) -- if you will use your system reply, use byte 16, if you counted 17 from 1 not 0


if midiChannelTxByte == 00 then
	midiChannelTx = 1
 	midiChannelTxCount = 17 -- to stop the timer loop (17> midiChannelTxCount limit) -- timer will reset it to 0 anyway
elseif midiChannelTxByte == 1 then
	midiChannelTx = 16
	midiChannelTxCount = 17 -- to stop the timer loop (17> midiChannelTxCount limit) -- timer will reset it to 0 anyway
...
else
	midiChannelTx = 17
	midiChannelTxCount = 17 -- to stop the timer loop (17> midiChannelTxCount limit) -- timer will reset it to 0 anyway
...
end


console("midiChannelTx is : "..(midiChannelTx))

```
now you've established a variable midiChannelTx

you can run other scripts based on this like : 
```
if midiChannelTx == 1 then 
-- do this
elseif midiChannelTx == 00 then
-- do that
end
```

This is a very quick strategy but I tried to draw the basis.

**Reply by Youngwhisk** [reply]
Thanks very much. I will not be able to test for a couple of days. These are all great ideas!!

**Reply by Youngwhisk** [reply]
Damien,
Thanks for this. I have finally gotten this to a state where it works. When the button is clicked, it steps through the channels and sends the dump request. If a midi response is received the timer stops and the correct channel is indicated in a console statement. I will be able to act on the result to do what I need.

The only issue I have is that when the **code is compiled**, The uiButton does nothing on the first click but performs properly on all clicks after that. I am not sure why. Anyone have any ideas?

Regarding the Midi Identity Request, all my midi devices are old (80's and 90's) and do not respond to that. (not mentioned in the midi implementation info either). I have an original Alesis HR-16 that supposedly has its own identity request but it doesn't work with mine. Mine does not have the menu toggle to disable or enable sysex. I think mine is a previous version before the ID request was implemented.

Thanks

**Reply by damiensellier** [reply]
That sounds good, it's a nice progress! Nevermind for the global sysex ID request, go for the system dump, just to get your reply.

When you load your panel the buttons sometimes switch from Off to On, so your first click could switch it back to Off which is what you don't want.
For your button initial click that does not work,  you have to set 2 values for the 2 positions of your button in its properties "component > button values"  (in panel mode) :
 
```
False = 0
True = 1
```

it could be whatever value you need for your message, for exemple : 

```
0
127
```
or 

```
Bypass Off = 0
Bypass On = 127
```

And you can add a default position on startup in your "panelLoaded" script (defined in the global properties panel) that sets it before your first click:

```
-- Set button init position to OFF before load
panel:getModulatorByName("NAMEOFYOURBUTTON"):setModulatorValue(0, false, false, false)
```

It should do the trick because your first click will go to the "on" or the 2nd value of the list, straight away :)
